tendermint/src/rpc/event_listener.rs

```diff
@@ -0,0 +1,292 @@
+    value: String,
+}
+
+///Block Results JSON PRC response
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct JSONRpcBlockResult {
+    jsonrpc: String,
+    id: String,
+    result: BlockResult,
+}
+/// Block Results
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct BlockResult {
+    query: String,
+    data: BlockResultData,
+    events: HashMap<String, Vec<String>>,
+}
+/// Block Results data
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct BlockResultData {
+    #[serde(rename = "type")]
+    data_type: String,
+    value: BlockValue,
+}
+///Block Value
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct BlockValue {
+    block: Block,
+    result_begin_block: ResultBeginBlock,
+    result_end_block: ResultEndBlock,
+}
+/// Block
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct Block {
+    header: Header,
+    data: BlockData,
+    evidence: Evidence,
+    last_commit: LastCommit,
+}
+///Block Txs
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct BlockData {
+    txs: Option<serde_json::Value>,
+}
+///Tendermint evidence
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct Evidence {
+    evidence: Option<serde_json::Value>,
+}
+/// Header
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct Header {
+    version: Version,
+    chain_id: String,
+    height: String,
+    time: String,
+    last_block_id: BlockId,
+    last_commit_hash: String,
+    data_hash: String,
+    validators_hash: String,
+    next_validators_hash: String,
+    consensus_hash: String,
+    app_hash: String,
+    last_results_hash: String,
+    evidence_hash: String,
+    proposer_address: String,
+}
+///Block ID
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct BlockId {
+    hash: String,
+    parts: Parts,
+}
+/// Block Parts
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct Parts {
+    total: String,
+    hash: String,
+}
+///Block version
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct Version {
+    block: String,
+    app: String,
+}
+///Perevious Commit
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct LastCommit {
+    height: String,
+    round: String,
+    block_id: BlockId,
+    signatures: Vec<Signature>,
+}
+///Signatures
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct Signature {
+    block_id_flag: i64,
+    validator_address: String,
+    timestamp: String,
+    signature: String,
+}
+/// Begin Blocke Envts
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct ResultBeginBlock {
+    events: Vec<TmEvent>,
+}
+///End Block Events
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct ResultEndBlock {
+    validator_updates: Vec<Option<serde_json::Value>>,
+}
+
+impl JSONRpcTxResult {
+    /// Extract events from TXEvent if event matches are type query
+    pub fn extract_events(
+        &self,
+        action_query: &str,
+    ) -> Result<HashMap<String, Vec<String>>, &'static str> {
```

Comment:
I'll make this a 
```suggestion
    ) -> Result<HashMap<String, Vec<String>>, Box<dyn std::Error>> {
```
tomorrow. (Do not commit yet).

Reply:
Yeah an actual error type or Box Dyn is one of the last things that needs to be done

Reply:
Did you want to sort this out? i'm down for a break now

Reply:
done in  668696d